File: benchmarks/src/lua/number_parse.lua
local n=200000;local seed=12345;local function ri(m)seed=(seed*16807)%2147483647;return seed%m end
local strings={};for _=1,n do local s="";if ri(2)==0 then s="-" end;local nd=1+ri(9);s=s..tostring(1+ri(9));for _=2,nd do s=s..tostring(ri(10))end;strings[#strings+1]=s end
local start=os.clock();local total=0;local M=1000003
for _,s in ipairs(strings)do local neg=false;local idx=1;if s:sub(1,1)=="-" then neg=true;idx=2 end;local val=0
while idx<=#s do local c=s:sub(idx,idx);local d;if c=="0"then d=0 elseif c=="1"then d=1 elseif c=="2"then d=2 elseif c=="3"then d=3 elseif c=="4"then d=4 elseif c=="5"then d=5 elseif c=="6"then d=6 elseif c=="7"then d=7 elseif c=="8"then d=8 else d=9 end
val=val*10+d;idx=idx+1 end;if neg then val=-val end;total=total+val end
local r=((total%M)+M)%M;local finish=os.clock();print("Result: "..r);print(string.format("Time: %.0fms",(finish-start)*1000))

File: benchmarks/src/perl/number_parse.pl
use Time::HiRes qw(time);
my $n = 200000; my $seed = 12345;
sub ri { my $m = shift; $seed = ($seed * 16807) % 2147483647; return $seed % $m; }
my @strings;
for (1..$n) {
    my $s = ""; $s = "-" if ri(2) == 0;
    my $nd = 1 + ri(9); $s .= (1 + ri(9));
    for (2..$nd) { $s .= ri(10); }
    push @strings, $s;
}
my $start = time(); my $total = 0; my $M = 1000003;
for my $s (@strings) {
    my $neg = 0; my $idx = 0;
    if (substr($s, 0, 1) eq '-') { $neg = 1; $idx = 1; }
    my $val = 0;
    while ($idx < length($s)) {
        my $c = substr($s, $idx, 1); my $d;
        if ($c eq '0') { $d = 0; } elsif ($c eq '1') { $d = 1; }
        elsif ($c eq '2') { $d = 2; } elsif ($c eq '3') { $d = 3; }
        elsif ($c eq '4') { $d = 4; } elsif ($c eq '5') { $d = 5; }
        elsif ($c eq '6') { $d = 6; } elsif ($c eq '7') { $d = 7; }
        elsif ($c eq '8') { $d = 8; } else { $d = 9; }
        $val = $val * 10 + $d; $idx++;
    }
    $val = -$val if $neg; $total += $val;
}
my $r = (($total % $M) + $M) % $M;
my $end = time(); my $ms = ($end - $start) * 1000;
print "Result: $r\n"; printf "Time: %.0fms\n", $ms;
